sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -119,7 +119,30 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       qualifiedTableName.database, qualifiedTableName.name)
 
     if (DDLUtils.isDatasourceTable(table)) {
-      val dataSourceTable = cachedDataSourceTables(qualifiedTableName)
+      val dataSourceTable =
+        cachedDataSourceTables(qualifiedTableName) match {
+          case l @ LogicalRelation(relation: HadoopFsRelation, _, _) =>
+            // Ignore the scheme difference when comparing the paths
+            val isSamePath =
+              table.storage.locationUri.isDefined && relation.location.rootPaths.size == 1 &&
+                table.storage.locationUri.get == relation.location.rootPaths.head.toUri.getPath
+            // If we have the same paths, same schema, and same partition spec,
+            // we will use the cached relation.
+            val useCached =
+              isSamePath &&
+              l.schema == table.schema &&
+              relation.bucketSpec == table.bucketSpec &&
+              relation.partitionSchema == table.partitionSchema
+            if (useCached) {
+              l
+            } else {
+              // If the cached relation is not updated, we invalidate it right away.
+              cachedDataSourceTables.invalidate(qualifiedTableName)
+              // Reload it from the external catalog
+              cachedDataSourceTables(qualifiedTableName)
+            }
+          case o => o
+        }
```

Comment:
The above fix is for ensuring the fix is small enough to be backported to the 2.1 branch. 

In the master branch, I tried to combine it with the existing [`getCached`](https://github.com/gatorsmile/spark/blob/d7cd667fa7ffdcccb1470c1e2ef6087d5e60f6b3/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala#L164-L212). However, it requires refactoring. Thus, it will be submitted as a separate PR. Thanks!